cpp/arcticdb/version/symbol_list.cpp

```diff
@@ -6,292 +6,738 @@
-                }
-            } else {
-                SYMBOL_LIST_RUNTIME_LOG("Not compacting the symbol list due to lock contention");
-            }
-        } catch (const storage::PermissionException& ex) {
-            // Note: this only reflects AN's permission check and is not thrown by the Storage
-            SYMBOL_LIST_RUNTIME_LOG("Not compacting the symbol list due to lack of permission");
-        } catch (const std::exception& ex) {
-            log::version().warn("Ignoring the error while trying to compact the symbol list: {}", ex.what());
+std::vector<SymbolListEntry> load_previous_from_version_keys(
+        const std::shared_ptr<Store>& store,
+        SymbolListData& data) {
+    std::vector<StreamId> stream_ids;
+    store->iterate_type(KeyType::VERSION_REF, [&data, &stream_ids] (const auto& key) {
+        auto id = variant_key_id(key);
+        stream_ids.push_back(id);
+
+        if (stream_ids.size() == data.max_delta_ * 2 && !data.warned_expected_slowdown_) {
+            log::symbol().warn(
+                "No compacted symbol list cache found. "
+                "`list_symbols` may take longer than expected. \n\n"
+                "See here for more information: https://docs.arcticdb.io/technical/on_disk_storage/#symbol-list-caching\n\n"
+                "To resolve, run `list_symbols` through to completion to compact the symbol list cache.\n"
+                "Note: This warning will only appear once.\n");
+
+            data.warned_expected_slowdown_ = true;
+        }
+    });
+    auto res = folly::collect(batch_get_latest_undeleted_and_latest_versions_async(store, data.version_map_, stream_ids)).get();
+
+    std::vector<SymbolListEntry> symbols;
+    for(auto&& [idx, opt_key_pair]: folly::enumerate(res)) {
+        const auto& [maybe_undeleted, _]  = opt_key_pair;
+        if(maybe_undeleted) {
+            const auto version_id = maybe_undeleted->version_id();
+            const auto timestamp = maybe_undeleted->creation_ts();
+            symbols.emplace_back(stream_ids[idx], version_id, timestamp, ActionType::ADD);
         }
     }
 
-    return load_result.detach_symbols();
+    data.version_map_->flush();
+    return symbols;
 }
 
-bool SymbolList::can_update_symbol_list(const std::shared_ptr<Store>& store,
-        const std::optional<KeyVectorItr>& maybe_last_compaction) {
-    bool found_last = false;
-    bool has_newer = false;
+std::vector<AtomKey> get_all_symbol_list_keys(
+        const std::shared_ptr<StreamSource>& store,
+        SymbolListData& data) {
+    std::vector<AtomKey> output;
+    uint64_t uncompacted_keys_found = 0;
+    store->iterate_type(KeyType::SYMBOL_LIST, [&data, &output, &uncompacted_keys_found] (auto&& key) {
+        auto atom_key = to_atom(std::forward<decltype(key)>(key));
+        if(atom_key.id() != compaction_id) {
+            uncompacted_keys_found++;
+        }
+        if (uncompacted_keys_found == data.max_delta_ * 2 && !data.warned_expected_slowdown_) {
```

Comment:
What is the idea of this check `uncompacted_keys_found == data.max_delta_ * 2`? Is this some sort of heuristic?